### PR TITLE
core(view): warn about the implications of without initializing

### PR DESCRIPTION
**Pull request by romintomasetti**
This PR describes what `Kokkos::WithoutInitialization` implies.

### CODE REVIEW

File: docs/Makefile [hunk 1]
```diff
@@ -1,5 +1,5 @@
 html:
-	sphinx-build -b html ./source/ ./generated_docs/ -W --keep-going
+	sphinx-build -W -b html ./source/ ./generated_docs/ -W --keep-going
```
Comment (romintomasetti) [inline docs/Makefile:2]: Probably good to extract this in a separate PR.

File: docs/source/ProgrammingGuide/View.rst [hunk 2]
```diff
@@ -0,0 +1,723 @@
+.. role:: cppcode(code)
```
Comment (romintomasetti) [inline docs/source/ProgrammingGuide/View.rst:1]: I appears like I've deleted the file and re-created it, but I think it's because there are so many changes that `git` thinks that's what I've done.

**Review by dalg24 — DISMISSED** [review]
Please do not mix the change of format with editing the actual content in the same PR

**Review by masterleinad — CHANGES_REQUESTED** [review]
Please split this pull request up. Either add the changes to initialization first or convert the file first. Also adding another flag in `doc/Makefile` should be a separate pull request.

File: docs/Makefile (hunk 1 above)
Comment (romintomasetti) [reply docs/Makefile:2]: See #570.

File: docs/source/ProgrammingGuide/View.rst (hunk 2 above)
Comment (romintomasetti) [reply docs/source/ProgrammingGuide/View.rst]: Outdated.

File: docs/source/ProgrammingGuide/View.rst [hunk 3]
```diff
@@ -458,6 +460,15 @@ in a way that reflects how first-touch affinity to threads is desired.
 Typically it is sufficient to use the parallel iteration index as the index of the location in the
 view to write to.
 
+.. warning::
+
+  :cppkokkos:`WithoutInitialization` implies that the destructor of each element of the :cppkokkos:`View` **will not be called**.
+  For instance, if you need the destructor of your :code:`MyObjectWithDestructor` be called for each element of your :cppkokkos:`View<MyObjectWithDestructor*>`,
+  you **should not use** :cppkokkos:`WithoutInitialization`.
```
Comment (masterleinad) [inline docs/source/ProgrammingGuide/View.rst:467]:
```suggestion
  For instance, if the View's value type is not trivially destructible,
  you **should not use** :cppkokkos:`WithoutInitialization` unless you are taking care of calling the destructor manually before the View deallocates its memory.
```

File: docs/source/ProgrammingGuide/View.rst [hunk 4]
```diff
@@ -458,6 +460,15 @@ in a way that reflects how first-touch affinity to threads is desired.
...
+
+  The reason for that is, since `Kokkos` did not initialize the memory, it cannot know onto which elements of the :cppkokkos:`View` calling the destructor
+  of the :code:`value_type` is valid.
```
Comment (masterleinad) [inline docs/source/ProgrammingGuide/View.rst:470]: Maybe, rather something like
```suggestion
The mental model is that whenever placement new is used to call the constructor, the destructor also isn't called before the memory is deallocated but it needs to be be called manually.
```
Or omit it with my previous suggestion.

File: docs/source/ProgrammingGuide/View.rst (hunk 3 above)
Comment (romintomasetti) [reply docs/source/ProgrammingGuide/View.rst:467]: Much better indeed.

File: docs/source/ProgrammingGuide/View.rst (hunk 4 above)
Comment (romintomasetti) [reply docs/source/ProgrammingGuide/View.rst:470]: :heavy_check_mark:

**Review by masterleinad — APPROVED** [review]

File: docs/source/ProgrammingGuide/View.rst [hunk 5]
```diff
@@ -458,6 +460,14 @@ in a way that reflects how first-touch affinity to threads is desired.
 Typically it is sufficient to use the parallel iteration index as the index of the location in the
 view to write to.
 
+.. warning::
+
+  :cppkokkos:`WithoutInitialization` implies that the destructor of each element of the :cppkokkos:`View` **will not be called**.
+  For instance, if the :cppkokkos:`View`'s value type is not trivially destructible,
+  you **should not use** :cppkokkos:`WithoutInitialization` unless you are taking care of calling the destructor manually before the :cppkokkos:`View` deallocates its memory.
+
+  The mental model is that whenever placement new is used to call the constructor, the destructor also isn't called before the memory is deallocated but it needs to be called manually.
```
Comment (ajpowelsnl) [inline docs/source/ProgrammingGuide/View.rst:469]: **_Suggestion_**:  maybe spell out what to do in the case of both trivially and non-trivially destructible data types, maybe even including a couple of simple code-snippet examples.  Perhaps also mention a couple of typical use cases where someone would want to use a View `WithoutInitialization`.

The sentence on line 469 is not clear, and is a little redundant with the immediately preceding statements.  It may not be necessary if the previous verbiage were sharpened up, with spelled out use cases and an couple of simple examples.

**Review by ajpowelsnl — COMMENTED** [review]
Good instinct to improve this documentation.  Cleaning up the `WithoutInitialization` section is the more actionable suggestion.  My comment about "novice" is one for broader consideration.

File: docs/source/ProgrammingGuide/View.rst [hunk 6]
```diff
@@ -97,7 +97,9 @@ Finally, note that virtual functions are technically allowed, but calling them i
 6.2.3 Can I make a View of Views?
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-NOVICES: THE ANSWER FOR YOU IS "NO."  PLEASE SKIP THIS SECTION.  
+.. warning::
+
+  NOVICES: THE ANSWER FOR YOU IS "NO."  PLEASE SKIP THIS SECTION.
```
[the remark this answers is not in the document]
Comment (dalg24) [reply docs/source/ProgrammingGuide/View.rst:102]: This is beyond the scope of this PR.  The text is not being changed, just the typeset.

**Review by crtrott — APPROVED** [review]
I think this is sufficient. I you don't understand the difference between trivially destructible and not trivially destructible you should not use View of Views - i.e. you are a `novice` and should have skipped this section.
